cmd/charm/charmcmd/resource.go

```diff
@@ -196,6 +200,120 @@ func uploadDockerResource(p uploadResourceParams) (int, error) {
 	return rev, nil
 }
 
+func uploadExternalDockerResource(p uploadResourceParams, ref reference.Named) (int, error) {
+	digest, err := imageDigestForReference(p, ref)
+	if err != nil {
+		return 0, errgo.Mask(err)
+	}
+	rev, err := p.client.AddDockerResource(p.charmId, p.resourceName, ref.Name(), digest)
+	if err != nil {
+		return 0, errgo.Notef(err, "cannot add docker resource")
+	}
+	return rev, nil
+}
+
+func imageDigestForReference(p uploadResourceParams, ref reference.Named) (string, error) {
+	endpoint := registryEndpointForReference(ref)
+	path := reference.Path(ref)
+	reqModifier, err := registryAuthorizer(endpoint, path)
+	if err != nil {
+		return "", errgo.Mask(err)
+	}
+	resp, err := dockerRegistryDo("HEAD", endpoint+path+"/manifests/"+referenceTagOrDigest(ref), reqModifier)
+	if err != nil {
+		return "", errgo.Mask(err)
+	}
+	resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		return "", errgo.Newf("cannot get information on %q: %v", ref, resp.Status)
+	}
+	if v := resp.Header.Get("Docker-Distribution-Api-Version"); !strings.HasPrefix(v, "registry/2.") {
+		return "", errgo.Newf("incompatible registry version %q", v)
+	}
+	digest := resp.Header.Get("Docker-Content-Digest")
+	if digest == "" {
+		return "", errgo.Newf("no digest in response")
+	}
+	if ref, ok := ref.(reference.Canonical); ok {
+		// The image is referred to by a digest and that works,
+		// so we're all good.
+		return string(ref.Digest()), nil
+	}
+	// The image is referred to by a tag; we don't know for sure
+	// that it was uploaded as a v2 image, so we check to see
+	// that we can get information on the image using the digest
+	// we've just discovered.
+	resp, err = dockerRegistryDo("HEAD", endpoint+path+"/manifests/"+digest, reqModifier)
+	if err != nil {
+		return "", errgo.Mask(err)
+	}
+	resp.Body.Close()
+	if resp.StatusCode != http.StatusOK {
+		if resp.StatusCode == http.StatusNotFound {
+			return "", errgo.Newf("cannot find image by version 2 digest; perhaps it was uploaded as a version 1 manifest")
+		}
+		return "", errgo.Newf("cannot verify image digest: %v", resp.Status)
+	}
+	return digest, nil
+}
+
+func dockerRegistryDo(method, url string, reqModifier transport.RequestModifier) (*http.Response, error) {
+	req, err := http.NewRequest(method, url, nil)
+	if err != nil {
+		return nil, errgo.Mask(err)
+	}
+	req.Header.Set("Accept", "application/vnd.docker.distribution.manifest.v2+json")
```

Comment:
See above for why this might not be complete.

Reply:
done